[SR-9402] Emit a warning on duplicate target dependencies

sjavora opened the pull request:
~Implements https://bugs.swift.org/browse/SR-9402, which calls for SwiftPM to emit a warning when a target has duplicate dependencies.~

SPM was already partially checking for this, so this updates/expands that check. I'm not in love with `duplicateDetectionHash`, but it does what it needs to.

SDGGiesbrecht commented on Sources/PackageLoading/ManifestLoader.swift:
```diff
@@ -1027,6 +1034,10 @@ extension Basics.Diagnostic {
     }
+    
+    static func duplicateTargetDependency(dependencyName: String) -> Self {
+        .warning("duplicate target dependency named '\(dependencyName)'")
```
Including the target name in the message would help the user find it faster.

SDGGiesbrecht commented on Sources/PackageModel/Manifest/TargetDescription.swift:
```diff
@@ -22,7 +22,7 @@ public struct TargetDescription: Equatable, Codable {
 
     /// Represents a target's dependency on another entity.
-    public enum Dependency: Equatable {
+    public enum Dependency: Equatable, Hashable {
```
Relying on the synthesized `Hashable` like this will not catch the following sort of duplicate:

```swift
"a", // i.e. .byName(name: "a")
.target(name: "a")
```

On the other hand, comparing only the name will snag valid pairs, such as this:

```swift
.target(name: "a"),
.product(name: "a", package: "c")
```

infotrinerdis replied:
Does `byName` always represent a target? If so, it might be easiest to convert one to the other for the purposes of this comparison 🤔

sjavora replied:
Does `byName` always represent a target? If so, it might be easiest to convert one to the other for this comparison 🤔

SDGGiesbrecht replied:
No. It is short for either `.target(name: "a")` or `.product(name: "a", package: "a"`), or in pre‐5.2 tools-versions, `.product(name: "a", package: /* all */)`. Figuring it out either requires more elaborate logic, or performing the check later in the loading process after these sorts of connections have already been resolved.

sjavora replied:
> either requires more elaborate logic

Would this not just duplicate the resolution process that takes place later?

> or performing the check later in the loading process

This makes more sense to me.

What is your opinion on this?

SDGGiesbrecht commented on Sources/PackageModel/Manifest/PackageConditionDescription.swift:
```diff
@@ -9,7 +9,7 @@
 
 /// Represents a manifest condition.
-public struct PackageConditionDescription: Codable, Equatable {
+public struct PackageConditionDescription: Codable, Equatable, Hashable {
```
Do you think it is worth also warning of dependencies that can be combined?

For example warning about this:

```swift
.target(name: "a", condition: .when(platforms: [.macOS])),
.target(name: "a", condition: .when(platforms: [.linux]))
```

And suggesting this:

```swift
.target(name: "a", condition: .when(platforms: [.macOS, .linux]))
```

sjavora replied:
That seems like a stylistic choice to me. At least in Swift itself, this is not something that is enforced, e.g.:
```swift
switch foo {
    case .a:
        print("not c")
    case .b:
        print("not c")
    case .c:
        print("c!")
}
```
vs 
```swift
switch foo {
    case .a, .b:
        print("not c")
    case .c:
        print("c!")
}
```
There is no warning to combine the first two cases.

But maybe the package manager has different conventions?

SDGGiesbrecht replied:
I have no strong opinion either way. It just came up while I was thinking through the implications of comparing by name, and I wondered what your thoughts were.

SDGGiesbrecht reviewed (commented):
I am not the primary reviewer, but I left some comments you can think about while you wait for the real review.

sjavora commented:
Well, it turns out SPM already checks for duplicate target dependencies [here](https://github.com/apple/swift-package-manager/blob/5c4bef08836f8f30437932470e281b53f69cb709/Sources/PackageLoading/PackageBuilder.swift#L785)! But it checks only by name so the valid so this point actually applies to the current implementation:

> On the other hand, comparing only the name will snag valid pairs, such as this:
> 
> ```swift
> .target(name: "a"),
> .product(name: "a", package: "c")
> ```

I updated the implementation to handle that case and also added some explicit `.target` and `.product` to the test, since it also only had dependencies "by name".

tomerd commented on Sources/PackageLoading/PackageBuilder.swift:
```diff
@@ -1470,3 +1469,20 @@ extension PackageBuilder {
+
+private extension Target.Dependency {
+    
+    var duplicateDetectionHash: Int {
```
thanks @sjavora, how about returning a string instead of a hash? e.g. (may not compile):

```
var nameAndType: String {
switch self {
             case .target:
                return "target-\(self.name)"
             case .product:
                return "product-\(self.name)"
         }
}
```

sjavora replied:
Thanks @tomerd, compiles just fine - updated 👍.

tomerd commented:
@swift-ci please smoke test

tomerd commented:
thanks @sjavora
